### PR TITLE
refactor(impala): modernize the impala backend

### PR DESCRIPTION
**Pull request by cpcloud**
This PR removes HDFS integration from the impala backend and uses upstream
docker images for testing.

Use of HDFS in the impala backend was responsible for a most of its complexity.

Also move away from an ancient version of impala that is no longer maintained.

The primary motivation here is to reduce the scope of the Impala backend, and
to lower its maintenance by supporting fewer features.

Generally speaking, the primary APIs that are affected are **DDL-related**:
**existing `SELECT`-style queries should continue to work.**

Things that are no longer supported after this PR:

* Automatic generation of `LIKE PARQUET` in `CREATE TABLE` statements (this can
  still be done by users if they really need it)
* Supplying an HDFS connection to `ibis.impala.connect`
* Pandas interop that interacts with HDFS (writing a DataFrame to a CSV on HDFS
  for use in creating a table). If impala wants to eventually support more
  efficient pandas integration (e.g., direct inserts via the HS2 protocol) then
  we can consider it. Otherwise the maintenance burden of this integration is
  too high.
* Concurrent queries **via threading** are no longer supported. There appears to be deadlock happening. Processed-based concurrent queries work just fine.

BREAKING CHANGE: Direct HDFS integration is removed and support for ingesting pandas DataFrames directly is as well. The Impala backend still works with HDFS, but data in HDFS must be managed outside of ibis.

### CODE REVIEW

**Comment by cpcloud** [comment]
I need to straighten out the lockfile situation here. The scipy upper bound is causing issues, but I think I can deal with it.

**Comment by contang0** [comment]
Happy to see Impala receiving attention, even if in this case it means losing one feature (create table). :-) When I started using Ibis it was indeed difficult to figure out how to set up the HDFS connection, which in the end turned out to not even be necessary for 99% of the things I wanted to do with Ibis.

BTW, are cross-schema joins coming to Impala backend? That's one thing that is difficult to live without. For now I rely on hacky workarounds.

**Comment by cpcloud** [comment]
@contang0 Thanks for chiming in! We really do depend on users like yourself to help us understand what to prioritize for backends like Impala.

Regarding `CREATE TABLE`, it's only `CREATE TABLE` with `LIKE PARQUET` support that is removed.

`CREATE TABLE AS SELECT ...` and creating tables from DataFrames should still work. The latter will be slower due to not writing the data to HDFS directly and instead generating an `INSERT` statement with the data.

**Comment by cpcloud** [comment]
> Also, are cross-schema joins coming to Impala backend?

I can't promise anything concrete, but we can certainly look into it!

**Comment by gforsyth** [comment]
Hey @cpcloud -- I lost track of this one.

Did this get oddly rebased?  There are a bunch of commits that are unrelated to Impala, and I think a number of them have already landed in 8.x.x

**Comment by cpcloud** [comment]
@gforsyth Should be fixed now.

File: ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/microsecond/out.sql [hunk 1]
```diff
@@ -1 +1 @@
-extract(`i`, 'microsecond')
\ No newline at end of file
+extract(`i`, 'microsecond') % 1000000
```
Comment (gforsyth) [inline ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/microsecond/out.sql:1]: is this correct?

File: ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/millisecond/out.sql [hunk 2]
```diff
@@ -1 +1 @@
-extract(`i`, 'millisecond')
\ No newline at end of file
+extract(`i`, 'millisecond') % 1000
```
Comment (gforsyth) [inline ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/millisecond/out.sql:1]: is this correct?

**Review by gforsyth — COMMENTED** [review]
A few small things, mostly around keeping Impala stuff in the Impala backend, but also if it's going to be a ton of work we can get to it in a followup

File: ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/millisecond/out.sql (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/millisecond/out.sql:1]: Yeah, this was surprising to me too. The behavior of `extract` for micros and millis changed in Impala.

Comment (gforsyth) [reply ibis/backends/impala/tests/snapshots/test_value_exprs/test_timestamp_extract_field/millisecond/out.sql:1]: weeeeird.  ok!

File: ibis/backends/base/sql/registry/timestamp.py [hunk 3]
```diff
@@ -17,6 +17,22 @@ def extract_field_formatter(translator, op):
...
+
+
+def extract_microsecond(translator, op):
+    arg = translator.translate(op.args[0])
+
+    # This is pre-2.0 Impala-style, which did not used to support the
+    # SQL-99 format extract($FIELD from expr)
+    return f"extract({arg}, 'microsecond') % 1000000"
+
+
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/base/sql/registry/timestamp.py:35]: Impala doesn't actually have a registry 😬 

We should definitely give it one, but not in this PR :)

[the remark this answers is not in the document]
Comment (gforsyth) [reply ibis/backends/base/sql/registry/timestamp.py:35]: :fire: :-1: :fire: :sob: :fire:

**Review by gforsyth — APPROVED** [review]
I think this is good to go.  We can add a proper registry in a follow-up.

**Comment by cpcloud** [comment]
Sweet, sweet deletion.

**Comment by contang0** [comment]
I know that this change has not been released yet, however I wanted to try it out already and installed it from master. Unfortunately I am not able to connect anymore. If I understand it correctly, Ibis is attempting to create a temporary database in `_ensure_temp_db_exists`, which does not work.

```
---------------------------------------------------------------------------
HiveServer2Error                          Traceback (most recent call last)
c:\Users\username\Desktop\Projects\projname\src\pairing.py in line 1
----> 281 prod_mmsrdl = connect_ibis(schema_name="schema_name", env="prod")

File ~\Desktop\Projects\projname\src\utils.py:35, in connect_ibis(schema_name, env)
---> 35 return ibis.impala.connect(
     36     host=hosts[env],
     37     database=schema_name,
     38     port=443,
     39     use_ssl=True,
     40     auth_mechanism="PLAIN",
     41     user=os.getlogin(),
     42     password=pw,
     43     use_http_transport=True,
     44     http_path="cliservice",
     45 )

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\__init__.py:107, in __getattr__.<locals>.connect(*args, **kwargs)
    106 def connect(*args, **kwargs):
--> 107     return backend.connect(*args, **kwargs)

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\base\__init__.py:847, in BaseBackend.connect(self, *args, **kwargs)
    824 """Connect to the database.
    825 
    826 Parameters
   (...)
    844     An instance of the backend
    845 """
    846 new_backend = self.__class__(*args, **kwargs)
--> 847 new_backend.reconnect()
    848 return new_backend

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\base\__init__.py:862, in BaseBackend.reconnect(self)
    860 def reconnect(self) -> None:
    861     """Reconnect to the database already configured with connect."""
--> 862     self.do_connect(*self._con_args, **self._con_kwargs)

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:212, in Backend.do_connect(self, host, port, database, timeout, use_ssl, ca_cert, user, password, auth_mechanism, kerberos_service_name, pool_size, **params)
    210 self.con = con
    211 self.options = {}
--> 212 self._ensure_temp_db_exists()

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:668, in Backend._ensure_temp_db_exists(self)
    666 name, path = options.impala.temp_db, options.impala.temp_path
    667 if name not in self.list_databases():
--> 668     self.create_database(name, path=path, force=True)

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:307, in Backend.create_database(self, name, path, force)
    295 """Create a new Impala database.
    296 
    297 Parameters
   (...)
    304     Forcibly create the database
    305 """
    306 statement = CreateDatabase(name, path=path, can_exist=force)
--> 307 self._safe_exec_sql(statement)

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:276, in Backend._safe_exec_sql(self, *args, **kwargs)
    275 def _safe_exec_sql(self, *args, **kwargs):
--> 276     with self._safe_raw_sql(*args, **kwargs):
    277         pass

File c:\Mambaforge\envs\env_name\Lib\contextlib.py:137, in _GeneratorContextManager.__enter__(self)
    135 del self.args, self.kwds, self.func
    136 try:
--> 137     return next(self.gen)
    138 except StopIteration:
    139     raise RuntimeError("generator didn't yield") from None

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:272, in Backend._safe_raw_sql(self, query)
    270 if not isinstance(query, str):
    271     query = query.compile()
--> 272 with contextlib.closing(self.raw_sql(query)) as cur:
    273     yield cur

File c:\Mambaforge\envs\env_name\Lib\site-packages\ibis\backends\impala\__init__.py:252, in Backend.raw_sql(self, query)
    249     cursor._wait_to_finish()
    251     util.log(query)
--> 252     cursor.execute_async(query)
    254     cursor._wait_to_finish()
    255 except (Exception, KeyboardInterrupt):

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:388, in HiveServer2Cursor.execute_async(self, operation, parameters, configuration)
    383     op = self.session.execute(self._last_operation_string,
    384                               configuration,
    385                               run_async=True)
    386     self._last_operation = op
--> 388 self._execute_async(op)

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:407, in HiveServer2Cursor._execute_async(self, operation_fn)
    405 self._reset_state()
    406 self._debug_log_state()
--> 407 operation_fn()
    408 self._last_operation_active = True
    409 self._debug_log_state()

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:383, in HiveServer2Cursor.execute_async.<locals>.op()
    380 else:
    381     self._last_operation_string = operation
--> 383 op = self.session.execute(self._last_operation_string,
    384                           configuration,
    385                           run_async=True)
    386 self._last_operation = op

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:1227, in HS2Session.execute(self, statement, configuration, run_async)
   1219 req = TExecuteStatementReq(sessionHandle=self.handle,
   1220                            statement=statement,
   1221                            confOverlay=configuration,
   1222                            runAsync=run_async)
   1223 # Do not try to retry http requests.
   1224 # Read queries should be idempotent but most dml queries are not. Also retrying
   1225 # query execution from client could be expensive and so likely makes sense to do
   1226 # it if server is also aware of the retries.
-> 1227 return self._operation('ExecuteStatement', req, False)

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:1148, in ThriftRPC._operation(self, kind, request, retry_on_http_error)
   1147 def _operation(self, kind, request, retry_on_http_error=False):
-> 1148     resp = self._rpc(kind, request, retry_on_http_error)
   1149     return self._get_operation(resp.operationHandle)

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:1085, in ThriftRPC._rpc(self, func_name, request, retry_on_http_error)
   1083 response = self._execute(func_name, request, retry_on_http_error)
   1084 self._log_response(func_name, response)
-> 1085 err_if_rpc_not_ok(response)
   1086 return response

File c:\Mambaforge\envs\env_name\Lib\site-packages\impala\hiveserver2.py:781, in err_if_rpc_not_ok(resp)
    777 def err_if_rpc_not_ok(resp):
    778     if (resp.status.statusCode != TStatusCode.SUCCESS_STATUS and
    779             resp.status.statusCode != TStatusCode.SUCCESS_WITH_INFO_STATUS and
    780             resp.status.statusCode != TStatusCode.STILL_EXECUTING_STATUS):
--> 781         raise HiveServer2Error(resp.status.errorMessage)

HiveServer2Error: AuthorizationException: User 'username' does not have privileges to execute 'CREATE' on: __ibis_tmp
```

**Pull request merged**